File: face/geomajas-face-dojo/client/src/main/resources/geomajas/gfx/painter/FeaturePainter.js
/*
 * This is part of Geomajas, a GIS framework, http://www.geomajas.org/.
 *
 * Copyright 2008-2011 Geosparc nv, http://www.geosparc.com/, Belgium.
 *
 * The program is available in open source according to the GNU Affero
 * General Public License. All contributions in this program are covered
 * by the Geomajas Contributors License Agreement. For full licensing
 * details, see LICENSE.txt in the project root.
 */

dojo.provide("geomajas.gfx.painter.FeaturePainter");
dojo.require("geomajas._base");
dojo.require("geomajas.gfx.Painter");

dojo.declare("FeaturePainter", Painter, {

	/**
	 * @class Painter implementation for features. This painter always paints
	 * in worldspace!
	 * @author Jan De Moerloose & Pieter De Graef
	 *
	 * @constructor
	 * @extends Painter
	 */
	constructor : function (mapWidget, mapView) {
		this.mapWidget = mapWidget;
		this.transform = new WorldViewTransformation(mapView);
	},
	
	/**
	 * The actual painting function. Currently paints selected features only.
	 * @param feature A Paintable object.
	 * @param graphics A GraphicsContext object, responsible for actual drawing.
	 */
	paint : function (feature, graphics) {
		var geometry = feature.getGeometry();
		
		// selected features only
		if(!feature.isSelected()){
			return;
		}
		
		var id = feature.getSelectionId();
		// no updates, delete first to update !!! => PDG: why is this??? speed perhaps...
//		if(graphics.getElementById(id)){
//			return;
//		}
		
		log.info("FeaturePainter : painting feature "+feature.getId());
		// Style is shared reference, so we have to clone it before we can change it !
		var style = feature.getStyle().clone();
					
		if(geometry.declaredClass == geomajas.GeometryTypes.POLYGON){
			style.copyFrom(this.mapWidget.getPolygonSelectStyle());
		}
		if(geometry.declaredClass == geomajas.GeometryTypes.LINESTRING){
			style.copyFrom(this.mapWidget.getLineSelectStyle());
		}
		if(geometry.declaredClass == geomajas.GeometryTypes.POINT){
			style.copyFrom(this.mapWidget.getPointSelectStyle());
		}
		if(geometry.declaredClass == geomajas.GeometryTypes.MULTIPOLYGON){
			style.copyFrom(this.mapWidget.getPolygonSelectStyle());
		}
		if(geometry.declaredClass == geomajas.GeometryTypes.MULTILINESTRING){
			style.copyFrom(this.mapWidget.getLineSelectStyle());
		}

		var options = {
			id : id, 
			style: style,
			worldspace : true
		};
		
		// clipped and normal drawing is the same now ?
		var viewGeometry = this.transform.worldGeometryToPan(geometry);
		if(geometry.declaredClass == geomajas.GeometryTypes.POLYGON){
			graphics.drawPolygon (viewGeometry,options);
		} else if (geometry.declaredClass == geomajas.GeometryTypes.LINESTRING) {
			graphics.drawLine (viewGeometry,options);
		} else if (geometry.declaredClass == geomajas.GeometryTypes.MULTIPOLYGON) {
			graphics.drawPolygon (viewGeometry,options);
		} else if (geometry.declaredClass == geomajas.GeometryTypes.MULTILINESTRING) {
			graphics.drawLine (viewGeometry,options);
		} else if (geometry.type == geomajas.GeometryTypes.POINT) {
			options.styleId = feature.styleId;
			graphics.drawSymbol (viewGeometry,options);
		}
	},

	/**
	 * Deletes a features from the map. Also deletes the label.
	 */
	deleteShape : function (/*Feature*/feature, /*GraphicsContext*/graphics) {
		graphics.deleteShape(feature.getSelectionId());
	}
});

File: client/geomajas-client-smartgwt/client/src/main/java/org/geomajas/smartgwt/client/gfx/painter/FeaturePainter.java
/*
 * This is part of Geomajas, a GIS framework, http://www.geomajas.org/.
 *
 * Copyright 2008-2013 Geosparc nv, http://www.geosparc.com/, Belgium.
 *
 * The program is available in open source according to the GNU Affero
 * General Public License. All contributions in this program are covered
 * by the Geomajas Contributors License Agreement. For full licensing
 * details, see LICENSE.txt in the project root.
 */

package org.geomajas.smartgwt.client.gfx.painter;

import org.geomajas.configuration.FeatureStyleInfo;
import org.geomajas.geometry.Coordinate;
import org.geomajas.smartgwt.client.gfx.MapContext;
import org.geomajas.smartgwt.client.gfx.Paintable;
import org.geomajas.smartgwt.client.gfx.PaintableGroup;
import org.geomajas.smartgwt.client.gfx.Painter;
import org.geomajas.smartgwt.client.gfx.style.ShapeStyle;
import org.geomajas.smartgwt.client.map.feature.Feature;
import org.geomajas.smartgwt.client.spatial.WorldViewTransformer;
import org.geomajas.smartgwt.client.spatial.geometry.Geometry;
import org.geomajas.smartgwt.client.spatial.geometry.LineString;
import org.geomajas.smartgwt.client.spatial.geometry.MultiLineString;
import org.geomajas.smartgwt.client.spatial.geometry.MultiPolygon;
import org.geomajas.smartgwt.client.spatial.geometry.Polygon;

/**
 * Painter implementation for {@link org.geomajas.smartgwt.client.map.feature.Feature}s.
 * 
 * @author Joachim Van der Auwera
 */
public class FeaturePainter implements Painter {

	private ShapeStyle pointSelectStyle;

	private ShapeStyle lineSelectStyle;

	private ShapeStyle polygonSelectStyle;

	// -------------------------------------------------------------------------
	// Constructors:
	// -------------------------------------------------------------------------

	/** No-arguments constructor for GWT. */
	public FeaturePainter() {
	}

	// -------------------------------------------------------------------------
	// Painter implementation:
	// -------------------------------------------------------------------------

	/**
	 * Return the class-name of the type of object this painter can paint.
	 * 
	 * @return Return the class-name as a string.
	 */
	public String getPaintableClassName() {
		return Feature.class.getName();
	}

	/**
	 * The actual painting function. Draws the circles with the object's id.
	 * 
	 * @param paintable
	 *            A {@link org.geomajas.smartgwt.client.gfx.paintable.Text} object.
	 * @param group
	 *            The group where the object resides in (optional).
	 * @param context
	 *            A MapContext object, responsible for actual drawing.
	 */
	public void paint(Paintable paintable, Object group, MapContext context) {
		if (paintable != null) {
			Feature feature = (Feature) paintable;
			WorldViewTransformer worldViewTransformer = feature.getLayer().getMapModel().getMapView()
					.getWorldViewTransformer();
			Geometry geometry = worldViewTransformer.worldToPan(feature.getGeometry());
			ShapeStyle style = createStyleForFeature(feature);
			PaintableGroup selectionGroup = feature.getLayer().getSelectionGroup();
			context.getVectorContext().drawGroup(selectionGroup, feature);
			String name = feature.getLayer().getId() + "-" + feature.getId();

			switch (geometry.getLayerType()) {
				case LINESTRING:
					context.getVectorContext().drawLine(feature, name, (LineString) geometry, style);
					break;
				case MULTILINESTRING:
					MultiLineString mls = (MultiLineString) geometry;
					for (int i = 0; i < mls.getNumGeometries(); i++) {
						context.getVectorContext().drawLine(feature, name + "." + i,
								(LineString) mls.getGeometryN(i), style);
					}
					break;
				case POLYGON:
					context.getVectorContext().drawPolygon(feature, name, (Polygon) geometry, style);
					break;
				case MULTIPOLYGON:
					MultiPolygon mp = (MultiPolygon) geometry;
					for (int i = 0; i < mp.getNumGeometries(); i++) {
						context.getVectorContext().drawPolygon(feature, name + "." + i,
								(Polygon) mp.getGeometryN(i), style);
					}
					break;
				case POINT:
					if (hasImageSymbol(feature)) {
						context.getVectorContext().drawSymbol(feature, name, geometry.getCoordinate(), null,
								feature.getStyleId() + "-selection");
					} else {
						context.getVectorContext().drawSymbol(feature, name, geometry.getCoordinate(), style,
								feature.getStyleId());
					}
					break;
				case MULTIPOINT:
					Coordinate[] coordinates = geometry.getCoordinates();
					if (hasImageSymbol(feature)) {
						for (int i = 0; i < coordinates.length; i++) {
							context.getVectorContext().drawSymbol(feature, name + "." + i, coordinates[i], null,
									feature.getStyleId() + "-selection");
						}
					} else {
						for (int i = 0; i < coordinates.length; i++) {
							context.getVectorContext().drawSymbol(feature, name + "." + i, coordinates[i], style,
									feature.getStyleId());
						}
					}
					break;
				default:
					throw new IllegalStateException("Cannot draw feature with Geometry type " +
							geometry.getLayerType());
			}
		}
	}

	/**
	 * Delete a {@link Paintable} object from the given {@link MapContext}. It the object does not exist, nothing will
	 * be done.
	 * 
	 * @param paintable
	 *            The object to be painted.
	 * @param group
	 *            The group where the object resides in (optional).
	 * @param context
	 *            The context to paint on.
	 */
	public void deleteShape(Paintable paintable, Object group, MapContext context) {
		context.getVectorContext().deleteGroup(paintable);
	}

	// -------------------------------------------------------------------------
	// Getters and setters:
	// -------------------------------------------------------------------------

	public ShapeStyle getLineSelectStyle() {
		return lineSelectStyle;
	}

	public void setLineSelectStyle(ShapeStyle lineSelectStyle) {
		this.lineSelectStyle = lineSelectStyle;
	}

	public ShapeStyle getPointSelectStyle() {
		return pointSelectStyle;
	}

	public void setPointSelectStyle(ShapeStyle pointSelectStyle) {
		this.pointSelectStyle = pointSelectStyle;
	}

	public ShapeStyle getPolygonSelectStyle() {
		return polygonSelectStyle;
	}

	public void setPolygonSelectStyle(ShapeStyle polygonSelectStyle) {
		this.polygonSelectStyle = polygonSelectStyle;
	}

	// Private methods:

	private ShapeStyle createStyleForFeature(Feature feature) {
		ShapeStyle style = null;
		if (null != feature) {
			FeatureStyleInfo styleInfo = null;
			if (feature.getStyleId() != null) {
				String check = feature.getStyleId();
				if (null != check) {
					for (FeatureStyleInfo fsi :
							feature.getLayer().getLayerInfo().getNamedStyleInfo().getFeatureStyles()) {
						if (check.equals(fsi.getStyleId())) {
							styleInfo = fsi;
							break;
						}
					}
				}
			}
			style = new ShapeStyle(styleInfo);
	
			Geometry geometry = feature.getGeometry();
			if (null != geometry) {
				switch (geometry.getLayerType()) {
					case LINESTRING:
					case MULTILINESTRING:
						style.merge(lineSelectStyle);
						break;
					case POLYGON:
					case MULTIPOLYGON:
						style.merge(polygonSelectStyle);
						break;
					case POINT:
					case MULTIPOINT:
						style.merge(pointSelectStyle);
						break;
					default:
						throw new IllegalStateException("Cannot create style for feature with Geometry type " +
								geometry.getLayerType());
				}
			}
		}
		return style;
	}

	private boolean hasImageSymbol(Feature feature) {
		String styleId = feature.getStyleId();
		for (FeatureStyleInfo style : feature.getLayer().getLayerInfo().getNamedStyleInfo().getFeatureStyles()) {
			if (style.getStyleId().equals(styleId)) {
				return style.getSymbol() != null && style.getSymbol().getImage() != null;
			}
		}
		return false;
	}
}
